synapse/handlers/federation.py

```diff
@@ -1649,3 +1640,55 @@ def get_next(it, opt=None):
             },
             "missing": [e.event_id for e in missing_locals],
         })
+
+    @defer.inlineCallbacks
+    @log_function
+    def exchange_third_party_invite(self, invite):
+        sender = invite["sender"]
+        room_id = invite["room_id"]
+
+        event_dict = {
+            "type": EventTypes.Member,
+            "content": {
+                "membership": Membership.INVITE,
+                "third_party_invite": invite,
+                "token": invite["token"],
+            },
+            "room_id": room_id,
+            "sender": sender,
+            "state_key": invite["mxid"],
+        }
+
+        if (yield self.auth.check_host_in_room(room_id, self.hs.hostname)):
+            builder = self.event_builder_factory.new(event_dict)
+            EventValidator().validate_new(builder)
+            event, context = yield self._create_new_client_event(builder=builder)
+            check = self.auth.check(event, context.current_state)
+            yield check()
+            member_handler = self.hs.get_handlers().room_member_handler
+            yield member_handler.change_membership(event, context)
+        else:
+            destinations = list(set([x.split(":", 1)[-1] for x in (sender, room_id)]))
```

Comment:
While that *does* work, the replication layer assumes that the iterable it's passed is ordered, and that the first element will be the first destination which is contacted.